lldb/source/Target/Target.cpp

```diff
@@ -2789,42 +2792,55 @@ llvm::Optional<SwiftScratchContextReader> Target::GetSwiftScratchContext(
     auto typesystem_sp = std::make_shared<TypeSystemSwiftTypeRefForExpressions>(
         lldb::eLanguageTypeSwift, *this, *lldb_module);
     typesystem_sp->GetSwiftASTContext();
-    m_scratch_typesystem_for_module.insert({idx, typesystem_sp});
+    m_scratch_typesystem_for_module.insert({key, typesystem_sp});
     if (log)
       log->Printf("created module-wide scratch context\n");
-    return typesystem_sp.get();
+    return;
   };
 
-  auto *swift_scratch_ctx = get_or_create_fallback_context();
-  if (!swift_scratch_ctx) {
-    if (log)
-      log->Printf("returned project-wide scratch context\n");
+  llvm::Optional<SwiftScratchContextReader> reader;
+  if (lldb_module && m_use_scratch_typesystem_per_module) {
+    maybe_create_fallback_context();
```

Comment:
Could you explain one more time why this can't potentially delete the scratch context while it's being used?

In my mind the following events could happen:

- Thread 1 successfully calls `GetSwiftScratchContext` and creates a new scratch type system.
- Thread 2 calls `GetSwiftScratchContext`
  - Thread 2 calls `maybe_create_fallback_context` (this call is not guarded by any locks as far as I can tell).
    - `maybe_create_fallback_context` deletes the scratch type system.
- Thread 1 is now using a dangling reference to the scratch type system.

Reply:
You are right.

Reply:
I addressed this by only deleting when we can acquire the write lock.